### PR TITLE
[PyOV] String type support

### PR DESCRIPTION
**Pull request by jiwaszki**
### Details:
 - Adding string support to Python API.
 - Added creation of Tensors from Python list of string/float/integer elements. (general feature!)
 - Allow different numpy string types to be used `np.str_`/`np.bytes_`/`<>|U`/`<>|S` and Python types `str`/`bytes`.\
     - To read more about code characters, refer to numpy docs: https://numpy.org/doc/stable/reference/arrays.scalars.html
 - Return decoded string data from inference in `np.str_` based type.
 - Added automatic conversions between Python data and C++ data:
     - `bytes_data` which returns data in form of `np.bytes_` and allows to override data of Tensor (automatically converting the input),
     - `str_data` which returns data in form of `np.str_`  allows to override data of Tensor (also automatically converting the input),
 - Automatic decoding/encoding of string data is replacing a need to call `np.char.decode/encode` functions.
 - Extended `_data_dispatch` to handle new string types both arrays and scalar-like versions of them, added support for inputs in form of lists.
 - Extended internal common code to support string-based data.
 - Refactored common code into three main namespaces: `array_helpers`, `string_helpers` and `type_helpers`
 - Added testcases to check all described scenarios.

### Tickets:
 - *125294*

### CODE REVIEW

File: src/bindings/python/src/openvino/runtime/ie_api.py [hunk 1]
```diff
@@ -262,7 +262,7 @@ def create_infer_request(self) -> InferRequest:
         """
         return InferRequest(super().create_infer_request())
 
-    def infer_new_request(self, inputs: Union[dict, list, tuple, Tensor, np.ndarray] = None) -> OVDict:
+    def infer_new_request(self, inputs: Any = None) -> OVDict:
```
Comment (p-wysocki) [inline src/bindings/python/src/openvino/runtime/ie_api.py:265]: Should we have `Any`? It doesn't say much, and now there's a discrepancy between Python typing and the docstring.

File: src/bindings/python/src/pyopenvino/core/common.cpp [hunk 2]
```diff
@@ -76,16 +106,121 @@ const TensorIndexMap cast_to_tensor_index_map(const py::dict& inputs) {
...
+        OPENVINO_THROW("Tensor's type must be a string!");
+    }
+    // Slow approach ~x5:
+    // py::list _list;
+    // for(size_t i = 0; i < t.get_size(); ++i) {
+    //     PyObject* _unicode_obj = PyBytes_FromStringAndSize(&data[i][0], data[i].length());
+    //     _list.append(_unicode_obj);
+    //     Py_XDECREF(_unicode_obj);
+    // }
+    // return py::array(_list);
```
Comment (p-wysocki) [inline src/bindings/python/src/pyopenvino/core/common.cpp:122]: Should we keep this in the codebase? From what I understand a faster method is implemented, is there a reason to have the slower approach around?

File: src/bindings/python/src/pyopenvino/core/common.cpp [hunk 3]
```diff
@@ -134,9 +269,18 @@ py::array as_contiguous(py::array& array, ov::element::Type type) {
 }
 
 py::array array_from_tensor(ov::Tensor&& t, bool is_shared) {
+    // Special case for string data type.
+    if (t.get_element_type() == ov::element::string) {
+        PyErr_WarnEx(PyExc_RuntimeWarning,
+                     "Data of string type will be copied! Please use dedicated functions "
```
Comment (p-wysocki) [inline src/bindings/python/src/pyopenvino/core/common.cpp:275]:
```suggestion
                     "Data of string type will be copied! Please use dedicated properties "
```

**Review by p-wysocki — COMMENTED** [review]
Can we get a very short note in some docstring or an inline comment about ctypes such as `|S8` and `<U`? It might be hard to find information on it, googling "string ctypes" doesn't help a lot. Maybe a link to some relevant source?

File: src/bindings/python/src/openvino/runtime/ie_api.py (hunk 1 above)
Comment (jiwaszki) [reply src/bindings/python/src/openvino/runtime/ie_api.py:265]: I will change a docstring then. It is the same as for all other infer functions -- just aligned them.

File: src/bindings/python/src/pyopenvino/core/common.cpp (hunk 2 above)
Comment (jiwaszki) [reply src/bindings/python/src/pyopenvino/core/common.cpp:122]: Leftover to be removed.

File: src/bindings/python/src/openvino/runtime/ie_api.py (hunk 1 above)
Comment (jiwaszki) [reply src/bindings/python/src/openvino/runtime/ie_api.py:265]: Done.

File: src/bindings/python/src/pyopenvino/core/common.cpp (hunk 2 above)
Comment (jiwaszki) [reply src/bindings/python/src/pyopenvino/core/common.cpp]: Done.

File: src/bindings/python/src/pyopenvino/core/common.cpp (hunk 3 above)
Comment (jiwaszki) [reply src/bindings/python/src/pyopenvino/core/common.cpp]: Done.

File: src/bindings/python/tests/test_runtime/test_tensor_string.py [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+        _ = ov.Tensor(data, shared_memory=True)
+    assert "SHARED MEMORY MODE FOR THIS TENSOR IS NOT APPLICABLE! String types can be only copied." in str(e.value)
+
+
+def test_string_tensor_data_warning():
+    data = np.array(["You", "shall", "not", "pass!"])
+    tensor = ov.Tensor(data, shared_memory=False)
+    with pytest.warns(RuntimeWarning) as w:
+        _ = tensor.data
+    assert "Data of string type will be copied! Please use dedicated functions" in str(w[0].message)
```
Comment (p-wysocki) [inline src/bindings/python/tests/test_runtime/test_tensor_string.py:51]:
```suggestion
    assert "Data of string type will be copied! Please use dedicated properties" in str(w[0].message)
```
Fix for tests

File: src/bindings/python/src/pyopenvino/core/common.cpp [hunk 5]
```diff
@@ -134,9 +257,18 @@ py::array as_contiguous(py::array& array, ov::element::Type type) {
 }
 
 py::array array_from_tensor(ov::Tensor&& t, bool is_shared) {
+    // Special case for string data type.
+    if (t.get_element_type() == ov::element::string) {
+        PyErr_WarnEx(PyExc_RuntimeWarning,
```
Comment (akuporos) [inline src/bindings/python/src/pyopenvino/core/common.cpp:262]: may be add a check if is_shared then throw warning?

**Review by akuporos — APPROVED** [review]

File: src/bindings/python/src/pyopenvino/core/common.cpp (hunk 5 above)
Comment (jiwaszki) [reply src/bindings/python/src/pyopenvino/core/common.cpp:262]: Actually, the goal is to redirect users to `bytes_str` and `string_str` to use it directly -- the behavior differs for strings and it's better to use them accordingly in different scenarios.

**Review by p-wysocki — APPROVED** [review]

**Comment by jiwaszki** [comment]
Awaiting the merge of https://github.com/openvinotoolkit/openvino/pull/21465 to resolve xfailed tests for data dispatcher.

**Review by akuporos — APPROVED** [review]

**Comment by akuporos** [comment]
Generally LGTM,
please improve docs a little bit in a separate PR: need to cover arguments, returns and their types

**Review by p-wysocki — APPROVED** [review]
LGTM regarding bugfixes and edge cases since my first review.

**Pull request merged**